Adds basic devcontainer for #1057

jacoverster opened the pull request:
Continuation of #1057. Adding a basic VSCode devcontainer with robust chrome and chromedriver installation.

All tests pass except those under tests/test_scene.py.

Next steps:
- Fix failing tests
- Consider multi-stage docker build for smaller images and faster build times

jacoverster commented:
Pytest output for debugging purposes:

`FAILED tests/test_scene.py::test_moving_sphere_with_timer - RuntimeError: Could not get position`
`FAILED tests/test_scene.py::test_no_object_duplication_on_index_client - assert 2 == 5`
`FAILED tests/test_scene.py::test_no_object_duplication_with_page_builder - assert 2 == 5`
`FAILED tests/test_scene.py::test_replace_scene - selenium.common.exceptions.JavascriptException: Message: javascript error: Cannot read properties of undefined (reading 'name')`
`FAILED tests/test_scene.py::test_create_dynamically - selenium.common.exceptions.NoSuchElementException: Message: no such element: Unable to locate element: {"method":"tag name","selector":"canvas"}`
`FAILED tests/test_scene.py::test_object_creation_via_context - selenium.common.exceptions.JavascriptException: Message: javascript error: Cannot read properties of undefined (reading 'name')`
`FAILED tests/test_scene.py::test_object_creation_via_attribute - selenium.common.exceptions.JavascriptException: Message: javascript error: Cannot read properties of undefined (reading 'name')`

rodja commented:
@jacoverster That is awesome. The scene also fails locally for us. It is an unresolved issue. See https://github.com/zauberzeug/nicegui/issues/643. I think you do not need to resolve it for the dev container to be included in the next release.

jacoverster commented:
Thanks @rodja, this is tested on Windows 11 with WSL2. Would be good if someone can test on other OS setups too and provide feedback.

natankeddem commented:
Okay, stupid question time. I cloned your repo, VS Code prompts to launch dev container, all good. Now I am in the container and feel like I am missing something basic. I should be able to launch the tests by executing `pytest` in terminal? I am just going by documentation in `CONTRIBUTING.md`.
```
$ ls    
CITATION.cff        DEPENDENCIES.md  deploy.sh               docker-entrypoint.sh  fetch_google_fonts.py  fly.dockerfile  nicegui                 npm.py         pyproject.toml      set_scale.sh     tests
CODE_OF_CONDUCT.md  LICENSE          development.dockerfile  docker.sh             fetch_tailwind.py      fly.toml        nicegui.code-workspace  poetry.lock    release.dockerfile  setup.py         website
CONTRIBUTING.md     README.md        docker-compose.yml      examples              fly-entrypoint.sh      main.py         npm.json                prometheus.py  sceenshots          test_startup.sh
$ pytest
/bin/sh: 8: pytest: not found
```

jacoverster commented:
Hi @natankeddem, with the container the testing now also happens inside the poetry virtual environment. You need to run `poetry run pytest`. Sometimes the poetry venv can break due to manual installing etc, you can run 'poetry install' again to restore it.

rodja commented:
For me (on mac, M1) the devcontainer does not build:
<details>
  <summary>
Expand 

``` 
#0 11.77 The following packages have unmet dependencies:
#0 11.82  google-chrome-stable:amd64 
```

</summary>
  
```
[3326 ms] Start: Run: docker buildx build --load --build-arg BUILDKIT_INLINE_CACHE=1 -f /var/folders/23/69ym20yd7n1ffsskbddkk_gc0000gn/T/devcontainercli/container-features/0.50.0-1694061341139/Dockerfile-with-features -t vsc-nicegui-eabe06171a0264f46bf8e005bdb27cefb1fd0de0ff90392b028d7d6e0d57bcc9 --target dev_containers_target_stage --build-arg _DEV_CONTAINERS_BASE_IMAGE=dev_container_auto_added_stage_label /Users/rodja/Projects/nicegui
[+] Building 24.5s (8/13)                                                       
 => [internal] load build definition from Dockerfile-with-features         0.0s
 => => transferring dockerfile: 2.15kB                                     0.0s
 => [internal] load .dockerignore                                          0.0s
 => => transferring context: 93B                                           0.0s
 => [internal] load metadata for docker.io/library/python:3.8-slim         1.4s
 => [auth] library/python:pull token for registry-1.docker.io              0.0s
 => CACHED [dev_container_auto_added_stage_label 1/8] FROM docker.io/libr  0.0s
 => [internal] load build context                                          0.4s
 => => transferring context: 271.20kB                                      0.3s
 => [dev_container_auto_added_stage_label 2/8] RUN groupadd --gid 1000 v  11.0s
 => ERROR [dev_container_auto_added_stage_label 3/8] RUN wget -nv -O /tm  11.9s 
------                                                                          
 > [dev_container_auto_added_stage_label 3/8] RUN wget -nv -O /tmp/chrome.deb https://dl.google.com/linux/chrome/deb/pool/main/g/google-chrome-stable/google-chrome-stable_116.0.5845.96-1_amd64.deb     && apt-get update && apt-get install -y /tmp/chrome.deb     && rm /tmp/chrome.deb:
#0 9.226 2023-09-07 04:36:03 URL:https://dl.google.com/linux/chrome/deb/pool/main/g/google-chrome-stable/google-chrome-stable_116.0.5845.96-1_amd64.deb [96765132/96765132] -> "/tmp/chrome.deb" [1]                                                             
#0 9.348 Get:1 http://deb.debian.org/debian bookworm InRelease [151 kB]         
#0 9.450 Get:2 http://deb.debian.org/debian bookworm-updates InRelease [52.1 kB]
#0 9.499 Get:3 http://deb.debian.org/debian-security bookworm-security InRelease [48.0 kB]
#0 9.542 Get:4 http://deb.debian.org/debian bookworm/main arm64 Packages [8803 kB]
#0 10.35 Get:5 http://deb.debian.org/debian bookworm-updates/main arm64 Packages [4960 B]
#0 10.35 Get:6 http://deb.debian.org/debian-security bookworm-security/main arm64 Packages [58.1 kB]
#0 11.04 Fetched 9118 kB in 2s (5098 kB/s)
#0 11.04 Reading package lists...
#0 11.36 Reading package lists...
#0 11.66 Building dependency tree...
#0 11.74 Reading state information...
#0 11.77 Some packages could not be installed. This may mean that you have
#0 11.77 requested an impossible situation or if you are using the unstable
#0 11.77 distribution that some required packages have not yet been created
#0 11.77 or been moved out of Incoming.
#0 11.77 The following information may help to resolve the situation:
#0 11.77 
#0 11.77 The following packages have unmet dependencies:
#0 11.82  google-chrome-stable:amd64 : Depends: libasound2:amd64 (>= 1.0.17) but it is not installable
#0 11.82                               Depends: libatk-bridge2.0-0:amd64 (>= 2.5.3) but it is not installable
#0 11.82                               Depends: libatk1.0-0:amd64 (>= 2.2.0) but it is not installable
#0 11.82                               Depends: libatspi2.0-0:amd64 (>= 2.9.90) but it is not installable
#0 11.82                               Depends: libc6:amd64 (>= 2.17) but it is not installable
#0 11.82                               Depends: libcairo2:amd64 (>= 1.6.0) but it is not installable
#0 11.82                               Depends: libcups2:amd64 (>= 1.6.0) but it is not installable
#0 11.82                               Depends: libcurl3-gnutls:amd64 but it is not installable or
#0 11.82                                        libcurl3-nss:amd64 but it is not installable or
#0 11.82                                        libcurl4:amd64 but it is not installable or
#0 11.82                                        libcurl3:amd64 but it is not installable
#0 11.82                               Depends: libdbus-1-3:amd64 (>= 1.9.14) but it is not installable
#0 11.82                               Depends: libdrm2:amd64 (>= 2.4.75) but it is not installable
#0 11.82                               Depends: libexpat1:amd64 (>= 2.0.1) but it is not installable
#0 11.82                               Depends: libgbm1:amd64 (>= 17.1.0~rc2) but it is not installable
#0 11.82                               Depends: libglib2.0-0:amd64 (>= 2.39.4) but it is not installable
#0 11.82                               Depends: libgtk-3-0:amd64 (>= 3.9.10) but it is not installable or
#0 11.82                                        libgtk-4-1:amd64 but it is not installable
#0 11.82                               Depends: libnspr4:amd64 (>= 2:4.9-2~) but it is not installable
#0 11.82                               Depends: libnss3:amd64 (>= 2:3.35) but it is not installable
#0 11.82                               Depends: libpango-1.0-0:amd64 (>= 1.14.0) but it is not installable
#0 11.82                               Depends: libvulkan1:amd64 but it is not installable
#0 11.82                               Depends: libx11-6:amd64 (>= 2:1.4.99.1) but it is not installable
#0 11.82                               Depends: libxcb1:amd64 (>= 1.9.2) but it is not installable
#0 11.82                               Depends: libxcomposite1:amd64 (>= 1:0.4.4-1) but it is not installable
#0 11.82                               Depends: libxdamage1:amd64 (>= 1:1.1) but it is not installable
#0 11.82                               Depends: libxext6:amd64 but it is not installable
#0 11.82                               Depends: libxfixes3:amd64 but it is not installable
#0 11.82                               Depends: libxkbcommon0:amd64 (>= 0.5.0) but it is not installable
#0 11.82                               Depends: libxrandr2:amd64 but it is not installable
#0 11.83 E: Unable to correct problems, you have held broken packages.
------
ERROR: failed to solve: executor failed running [/bin/sh -c wget -nv -O /tmp/chrome.deb https://dl.google.com/linux/chrome/deb/pool/main/g/google-chrome-stable/google-chrome-stable_${CHROME_VERSION}-1_amd64.deb     && apt-get update && apt-get install -y /tmp/chrome.deb     && rm /tmp/chrome.deb]: exit code: 100
[28200 ms] Error: Command failed: docker buildx build --load --build-arg BUILDKIT_INLINE_CACHE=1 -f /var/folders/23/69ym20yd7n1ffsskbddkk_gc0000gn/T/devcontainercli/container-features/0.50.0-1694061341139/Dockerfile-with-features -t vsc-nicegui-eabe06171a0264f46bf8e005bdb27cefb1fd0de0ff90392b028d7d6e0d57bcc9 --target dev_containers_target_stage --build-arg _DEV_CONTAINERS_BASE_IMAGE=dev_container_auto_added_stage_label /Users/rodja/Projects/nicegui
[28200 ms]     at IAA (/Users/rodja/.vscode/extensions/ms-vscode-remote.remote-containers-0.304.0/dist/spec-node/devContainersSpecCLI.js:463:1698)
[28200 ms]     at process.processTicksAndRejections (node:internal/process/task_queues:96:5)
[28200 ms]     at async qw (/Users/rodja/.vscode/extensions/ms-vscode-remote.remote-containers-0.304.0/dist/spec-node/devContainersSpecCLI.js:462:1691)
[28200 ms]     at async mK (/Users/rodja/.vscode/extensions/ms-vscode-remote.remote-containers-0.304.0/dist/spec-node/devContainersSpecCLI.js:462:610)
[28200 ms]     at async mAA (/Users/rodja/.vscode/extensions/ms-vscode-remote.remote-containers-0.304.0/dist/spec-node/devContainersSpecCLI.js:479:3660)
[28200 ms]     at async LC (/Users/rodja/.vscode/extensions/ms-vscode-remote.remote-containers-0.304.0/dist/spec-node/devContainersSpecCLI.js:479:4775)
[28201 ms]     at async jeA (/Users/rodja/.vscode/extensions/ms-vscode-remote.remote-containers-0.304.0/dist/spec-node/devContainersSpecCLI.js:611:12219)
[28201 ms]     at async _eA (/Users/rodja/.vscode/extensions/ms-vscode-remote.remote-containers-0.304.0/dist/spec-node/devContainersSpecCLI.js:611:11960)
[28203 ms] Exit code 1
[28205 ms] Command failed: /Applications/Visual Studio Code.app/Contents/Frameworks/Code Helper (Plugin).app/Contents/MacOS/Code Helper (Plugin) --ms-enable-electron-run-as-node /Users/rodja/.vscode/extensions/ms-vscode-remote.remote-containers-0.304.0/dist/spec-node/devContainersSpecCLI.js up --user-data-folder /Users/rodja/Library/Application Support/Code/User/globalStorage/ms-vscode-remote.remote-containers/data --container-session-data-folder /tmp/devcontainers-3487b177-3bf5-4e09-828c-156bd1953f3a1694061337465 --workspace-folder /Users/rodja/Projects/nicegui --workspace-mount-consistency cached --id-label devcontainer.local_folder=/Users/rodja/Projects/nicegui --id-label devcontainer.config_file=/Users/rodja/Projects/nicegui/.devcontainer/devcontainer.json --log-level debug --log-format json --config /Users/rodja/Projects/nicegui/.devcontainer/devcontainer.json --default-user-env-probe loginInteractiveShell --mount type=volume,source=vscode,target=/vscode,external=true --skip-post-create --update-remote-user-uid-default on --mount-workspace-git-root true
[28205 ms] Exit code 1
```
</details>

rodja commented:
>  with the container the testing now also happens inside the poetry virtual environment. You need to run poetry run pytest

@jacoverster Would it not be more convenient to istall it not in a virtual env? It's a container so installing it "globally" should be fine and would simplify usage.

rodja commented:
Seems like Google [Chrome is not available for Debian ARM processor yet](https://codetryout.com/google-chrome-following-packages-have-unmet-dependencies/#:~:text=Google%20Chrome%20is%20not%20available%20for%20Debian%20ARM%20processor%20yet%20(2023/May)). Because I run mac on an M1 chip chrome can not be installed.

rodja commented:
Maybe it's possible to switch to chromium-browser?

jacoverster commented:
> Maybe it's possible to switch to chromium-browser?

It looks like you [can access the architecture info during build](https://docs.docker.com/engine/reference/builder/#automatic-platform-args-in-the-global-scope) so we should be able to install processor specific packages.

jacoverster commented:
@rodja can you maybe modify your Dockerfile to a simplified version like this and populate the RUN section under install google chrome for ARM until it builds. Once it builds, add back the rest of the Dockerfile and check if the tests pass?

```
FROM python:3.8-slim

# Chrome and chromedriver versions must match as per
# https://googlechromelabs.github.io/chrome-for-testing/#stable
ARG CHROME_VERSION=116.0.5845.96

# Install google chrome for ARM
RUN ...

# Install chromedriver to a PATH directory
RUN wget -nv -O /tmp/chromedriver.zip https://edgedl.me.gvt1.com/edgedl/chrome/chrome-for-testing/${CHROME_VERSION}/linux64/chromedriver-linux64.zip \
    && unzip -j /tmp/chromedriver.zip -d /usr/local/bin/ \
    && rm /tmp/chromedriver.zip
```

jacoverster commented:
![image](https://github.com/zauberzeug/nicegui/assets/22006760/1c8a94bc-c146-4924-9875-65f423fd9407)

The chromedriver will probably also need to change. All the chrome and chromedriver build versions are available [here](https://googlechromelabs.github.io/chrome-for-testing/#stable) so maybe we could install it the same way?

rodja commented:
@jacoverster it seems like there is no chrome package available for ARM: https://stackoverflow.com/a/71174097/364388

jacoverster commented:
> It's a container so installing it "globally" should be fine and would simplify usage.

@rodja I tried following the best practices described in [this discussion](https://github.com/orgs/python-poetry/discussions/1879) on the poetry repo and summarized in this [blog post](https://medium.com/@albertazzir/blazing-fast-python-docker-builds-with-poetry-a78a66f5aed0) by one of the poetry contributors.

Would it be possible to completely remove poetry and just install from the pytproject.toml file using pip?

rodja commented:
I've done some experiments/modifications with the Dockerfile to use chromium instead of chrome:

```Dockerfile
FROM python:3.8-slim

# Chrome and chromedriver versions must match as per
# https://googlechromelabs.github.io/chrome-for-testing/#stable
ARG CHROME_VERSION=116.0.5845.96

ARG USERNAME=vscode
ARG USER_UID=1000
ARG USER_GID=$USER_UID

ENV POETRY_VERSION=1.6.1 \
    PATH="/home/$USERNAME/.local/bin:$PATH" \
    POETRY_NO_INTERACTION=1 \
    POETRY_VIRTUALENVS_IN_PROJECT=1 \
    POETRY_VIRTUALENVS_CREATE=1 \
    POETRY_CACHE_DIR=/tmp/poetry_cache \
    DISPLAY=:99 \
    DEBIAN_FRONTEND=noninteractive

# Create remote user
RUN groupadd --gid $USER_GID $USERNAME \
    && useradd --uid $USER_UID --gid $USER_GID -m $USERNAME \
    && apt-get update \
    && apt-get install -y sudo git wget gnupg unzip curl build-essential \
    && echo $USERNAME ALL=\(root\) NOPASSWD:ALL > /etc/sudoers.d/$USERNAME \
    && chmod 0440 /etc/sudoers.d/$USERNAME \
    && rm -rf /var/lib/apt/lists/*

# Install webbrowser and webdriver
RUN apt-get update \
    && apt-get install -y chromium chromium-driver \
    && rm -rf /var/lib/apt/lists/*

USER $USERNAME
WORKDIR /workspaces/nicegui

RUN pip install -U pip && pip install poetry==$POETRY_VERSION

COPY ./pyproject.toml ./poetry.lock* main.py ./
RUN poetry install --all-extras

CMD python -m debugpy --listen 5678 main.py
```

It starts up fine but the tests failed. The webdriver stuff in `conftest.py` might need to be changed. But I wont find the time to work on that in the next days.

jacoverster commented:
> I've done some experiments/modifications with the Dockerfile to use chromium instead of chrome

Ok great, I'll modify the Dockerfile to accommodate these changes towards being more platform-specific in the meantime.

jacoverster commented:
@natankeddem to disable the venv just modify the env variables:

```
    ...
    POETRY_VIRTUALENVS_IN_PROJECT=false \
    POETRY_VIRTUALENVS_CREATE=false \
    ...
```

And be sure to delete the .venv folder in your project directory, otherwise poetry will detect it and use it. I'll push an update for this shortly.

natankeddem commented:
> @natankeddem to disable the venv just modify the env variables:
> 
> ```
>     ...
>     POETRY_VIRTUALENVS_IN_PROJECT=false \
>     POETRY_VIRTUALENVS_CREATE=false \
>     ...
> ```
> 
> And be sure to delete the .venv folder in your project directory, otherwise poetry will detect it and use it. I'll push an update for this shortly.

This is giving me a permissions error:
```
=> ERROR [dev_container_auto_added_stage_label 8/8] RUN poetry install -  2.1s 
------                                                                          
 > [dev_container_auto_added_stage_label 8/8] RUN poetry install --all-extras:  
#0 0.912 Skipping virtualenv creation, as specified in config file.             
#0 1.015 Installing dependencies from lock file                                 
#0 1.300 
#0 1.300 Package operations: 89 installs, 10 updates, 0 removals
#0 1.300 
#0 1.301   • Updating attrs (23.1.0 /home/vscode/.local/lib/python3.8/site-packages -> 23.1.0)
#0 2.045 
#0 2.045   PermissionError
#0 2.045 
#0 2.045   [Errno 13] Permission denied: '/usr/local/lib/python3.8/site-packages/attr'
#0 2.045 
#0 2.045   at /usr/local/lib/python3.8/pathlib.py:1288 in mkdir
#0 2.087       1284│         """
#0 2.087       1285│         if self._closed:
#0 2.087       1286│             self._raise_closed()
#0 2.087       1287│         try:
#0 2.087     → 1288│             self._accessor.mkdir(self, mode)
#0 2.087       1289│         except FileNotFoundError:
#0 2.087       1290│             if not parents or self.parent == self:
#0 2.087       1291│                 raise
#0 2.087       1292│             self.parent.mkdir(parents=True, exist_ok=True)
#0 2.087 
------
Dockerfile-with-features:46
--------------------
  44 |     
  45 |     COPY ./pyproject.toml ./poetry.lock* main.py ./
  46 | >>> RUN poetry install --all-extras
  47 |     
  48 |     CMD python -m debugpy --listen 5678 main.py
```

jacoverster commented:
@rodja I noticed that the `pip install -e .` command is usually run outside the docker containers. Do you prefer this, or can I add it into the container? Alternatively we could add it as a postCreate command in the devcontainer.json file which will always run upon creation.

jacoverster commented:
> This is giving me a permissions error

Just have a look at the updated Dockerfile, the `USER $USERNAME` line had to be moved down to avoid this.

jacoverster commented:
Sorry @natankeddem I forgot to add the commit to the PR, will add now.

natankeddem commented:
> Sorry @natankeddem I forgot to add the commit to the PR, will add now.

Okay just let me know when things are in a state where you feel comfortable with me testing things out more.

jacoverster commented:
> Okay just let me know when things are in a state where you feel comfortable with me testing things out more

Please continue, I pushed the commit.

natankeddem commented:
Okay, container builds successfully now. I got dependency errors with `poetry run pytest` so I did a `poetry install` but that gave me permission errors. I haven't dug into things but shouldn't the `RUN poetry install --all-extras` in your `Dockerfile` be installing all the python dependencies? The user inside the container might not have permissions to install things after the container build process depending on how things are setup.

```
$ poetry run pytest
Skipping virtualenv creation, as specified in config file.
ImportError while loading conftest '/workspaces/nicegui/tests/conftest.py'.
tests/conftest.py:12: in <module>
    from nicegui import Client, binding, globals  # pylint: disable=redefined-builtin
nicegui/__init__.py:1: in <module>
    from . import elements, globals, ui  # pylint: disable=redefined-builtin
nicegui/ui.py:95: in <module>
    from .element import Element as element
nicegui/element.py:14: in <module>
    from .dependencies import Component, Library, register_library, register_vue_component
nicegui/dependencies.py:11: in <module>
    from .version import __version__
nicegui/version.py:3: in <module>
    __version__: str = importlib.metadata.version('nicegui')
E   importlib.metadata.PackageNotFoundError: nicegui
$ poetry install
Skipping virtualenv creation, as specified in config file.
Installing dependencies from lock file

Package operations: 0 installs, 0 updates, 14 removals

  • Removing bottle (0.12.25): Failed

  CalledProcessError

  Command '['/usr/local/bin/python', '-m', 'pip', 'uninstall', 'bottle', '-y']' returned non-zero exit status 2.

  at /usr/local/lib/python3.8/subprocess.py:516 in run
       512│             # We don't call process.wait() as .__exit__ does that for us.
       513│             raise
       514│         retcode = process.poll()
       515│         if check and retcode:
    →  516│             raise CalledProcessError(retcode, process.args,
       517│                                      output=stdout, stderr=stderr)
       518│     return CompletedProcess(process.args, retcode, stdout, stderr)
       519│ 
       520│ 

The following error occurred when trying to handle this error:


  EnvCommandError

  Command ['/usr/local/bin/python', '-m', 'pip', 'uninstall', 'bottle', '-y'] errored with the following return code 2
  
  Output:
  Found existing installation: bottle 0.12.25
  Uninstalling bottle-0.12.25:
  ERROR: Exception:
  Traceback (most recent call last):
    File "/usr/local/lib/python3.8/shutil.py", line 791, in move
      os.rename(src, real_dst)
  PermissionError: [Errno 13] Permission denied: '/usr/local/bin/bottle.py' -> '/tmp/pip-uninstall-bt4_xhdt/bottle.py'
...
...
```

jacoverster commented:
@natankeddem as per the current setup you first need to install nicegui locally with `pip install -e .`. I'm awaiting feedback to maybe put it into the container, but you will need to install it like that for now.

jacoverster commented:
I suggest we update the relevant contribution/test documentation after this has been thoroughly tested @rodja.

natankeddem commented:
Makes sense, I am all for removing additional steps if you can. To get this to work I had to make a venv manually then select it in the dev container then `pip install -e .`. I am not even sure if what I did was really the correct way of doing things but I got `pytest` to run. It would be great if you can just click the open dev container once VS Code loads then be off to developing without any additional thought required as to what dependencies to install and which tool to use (pip, poetry, ???).

jacoverster commented:
Yes @natankeddem, that was not needed. All the dependencies are there except local `nicegui`. I think it will be best to add a postCreateCommand to the devcontainer.json file. Then it will be ready to go after build 👍🏻 Let's just hear from @rodja if there is any reason not to do it.

natankeddem commented:
A few thoughts:
- Should the uid/gid ENV variable naming remain consistent with the existing `docker-entrypoint.sh`?
- Expose these and any other ENV variables in the `devcontainer.json` so people can easily change them? I think `remoteEnv` can be used for this? I might be wrong here...
- Tie the steps during the build which allow the ability to sudo to an ENV variable? Some people think the ability to sudo in a container is insecure, although I know there might be times when it might be useful.

rodja commented:
I think running `pip install -e .` inside the container is a great solution. It will make nicegui available as a package globally. So if developers uses the container and add their own project in /workspaces they will also be able to simply import NiceGUI. And when the core NiceGUI project is modified all changes will automatically be available to the other project which builds on NiceGUI. We use this technique intensively at Zauberzeug when developing RoSys and other projects.

jacoverster commented:
@rodja I've made some changes to try and add multi-platform support. The chrome install started cluttering the Dockerfile so I moved it into install_chrome.sh. This is nice for isolation and debugging, and it can easily be extended to include additional dependencies that might be needed for the other platfroms.

@natankeddem the `nicegui` local install now happens seamlessly inside the container, once it's built you can simply run `pytest` or even `poetry run pytest` out of the box.

**A note on having non-root and sudo inside the devcontainer:**
It's best-practice to not run as root user OR have sudo access in a container, but for a development container it is sometimes helpful to add new dependencies and for this you need to run `sudo poetry add ...` as you are running a non-root user. The long term goal is typically to have your dev container as close to the production one as possible, so once you move to production you simply remove sudo rights.

jacoverster commented:
> * Should the uid/gid ENV variable naming remain consistent with the existing `docker-entrypoint.sh`?

Will have to do som research on how to do this with a devcontainer. Should be possible.

> * Expose these and any other ENV variables in the `devcontainer.json` so people can easily change them? I think `remoteEnv` can be used for this? I might be wrong here...

Good idea, should be possible.

rodja commented:
I would rather like to move away from chrome in favour of chromium for the dev container on all platforms because:

- simplified maintainance
- same results on all platforms
- support for Mac M1 from the start

jacoverster commented:
> I would rather like to move away from chrome in favour of chromium for the dev container on all platforms because:
> 
> * simplified maintainance
> * same results on all platforms
> * support for Mac M1 from the start

Ok, makes sense, sorry I did not understand the nuance between chrome and chromium. I'll make the change to switch to chromium.

jacoverster commented:
I've updated to use Chromium and chromium driver, I think it is still important to have the versions matched. Tests pass on my side, please test on Mac @rodja?

rodja commented:
Thanks for making these changes @jacoverster. Looks good. I introduced a multi-stage build to compile netifaces which is not available as pre-build binary on MacOS ARM (M1).
Because the container is meant for development we could also simplify things by directly installing the gcc apt package into the dev container (and getting rid of the multi-stage build). What do you think?

rodja commented:
I was still not able to run the tests. Chromium seems unable to start:

```bash
vscode@docker-desktop:/workspaces/nicegui$ chromium --headless --no-sandbox --disable-gpu                             
find: ‘/home/vscode/.config/chromium/Crash Reports/pending/’: No such file or directory
[0920/043832.683728:ERROR:bus.cc(406)] Failed to connect to the bus: Failed to connect to socket /run/dbus/system_bus_socket: No such file or directory
[0920/043832.691145:ERROR:bus.cc(406)] Failed to connect to the bus: Failed to connect to socket /run/dbus/system_bus_socket: No such file or directory
[0920/043832.691487:ERROR:bus.cc(406)] Failed to connect to the bus: Failed to connect to socket /run/dbus/system_bus_socket: No such file or directory
[0920/043832.706530:WARNING:bluez_dbus_manager.cc(247)] Floss manager not present, cannot set Floss enable/disable.
```

But they are running fine on linux.

rodja commented:
Is anyone able to help with getting the tests to pass in dev container on MacOS ARM? I'm a bit busy at the moment...

rodja commented:
I've some other remarks/thoughts:

1. Could we change the default shell from `/bin/sh` to bash or zsh?
2. Is the  a way to allow `git push` from inside the container? Maybe in an other PR...
3. I don't think we need to specify CHROMIUM_VERSION in the Dockerfile. The Chomium dependencies are installed with apt which already ensures that the version will match. If we define a version here we will need to update the Dockerfile as soon as versions become unavailable -- which will lead to rebuilding devcontainers.
4. We will also need to update the CONTRIBUTING.md file to explain using the devcontainer.

rodja commented:
Regarding the tests on MacOS ARM (M1): I am able to take screenshots inside the dev container:

```bash
vscode@docker-desktop:/workspaces/nicegui$ chromium --headless --no-sandbox --disable-gpu --screenshot https://www.example.com
find: ‘/home/vscode/.config/chromium/Crash Reports/pending/’: No such file or directory
[0920/053347.232624:ERROR:bus.cc(406)] Failed to connect to the bus: Failed to connect to socket /run/dbus/system_bus_socket: No such file or directory
[0920/053347.239958:ERROR:bus.cc(406)] Failed to connect to the bus: Failed to connect to socket /run/dbus/system_bus_socket: No such file or directory
[0920/053347.240255:ERROR:bus.cc(406)] Failed to connect to the bus: Failed to connect to socket /run/dbus/system_bus_socket: No such file or directory
[0920/053347.251662:WARNING:bluez_dbus_manager.cc(247)] Floss manager not present, cannot set Floss enable/disable.
27083 bytes written to file screenshot.png
```

But starting with selenium fails:

```
>       raise exception_class(message, screen, stacktrace)
E       selenium.common.exceptions.SessionNotCreatedException: Message: session not created: Chrome failed to start: exited normally.
E         (chrome not reachable)
E         (The process started from chrome location /home/vscode/.cache/selenium/chrome/linux64/117.0.5938.88/chrome is no longer running, so ChromeDriver is assuming that Chrome has crashed.)
```

If I execute `/home/vscode/.cache/selenium/chrome/linux64/117.0.5938.88/chrome` I get

```
qemu-x86_64: Could not open '/lib64/ld-linux-x86-64.so.2': No such file or directory
```

which indicates that there's an issue with the emulation layer for running x86_64 binaries on the ARM architecture. But why? I have

```bash
vscode@docker-desktop:/workspaces/nicegui$ uname -m
aarch64
```

natankeddem commented:
@rodja does this help?

https://github.com/browserless/chrome/issues/2004#issuecomment-1453092994

jacoverster commented:
> 1. Could we change the default shell from `/bin/sh` to bash or zsh?

I've added a `SHELL` env variable for this, should force `bash` now.

> 2\. Is the  a way to allow `git push` from inside the container? Maybe in an other PR...

`git` is already installed in the container and works for me, what issues are you having with this?

> 3\. I don't think we need to specify CHROMIUM_VERSION in the Dockerfile. The Chomium dependencies are installed with apt which already ensures that the version will match. If we define a version here we will need to update the Dockerfile as soon as versions become unavailable -- which will lead to rebuilding devcontainers.

I've removed this, we must just be careful going forward as `chromium` and `chromium-driver` might also be on separate release cycles, they line up nicely atm but this might not be the case later.

> 5\. I would suggest to split the chomium package installation and user creation into seperate `RUN` commands in the Dockerfile to gain more clarity.

I will do this, makes sense.

jacoverster commented:
> Because the container is meant for development we could also simplify things by directly installing the gcc apt package into the dev container (and getting rid of the multi-stage build). What do you think?

Great! Thanks, I like the idea of moving to a multi-stage build to try and keep the devcontainer small. I'll see if I can move some other installs over to streamline the container.

rodja commented:
> Hi @rodja should we specify the version here otherwise it might cause conflict with the poetry.lock?

Yes, specifying the exact version would be good. But that would also nudge me even further in the direction of installing gcc directly in the dev container and getting rid of all the multi-step building etc. Reduction of complexity is very beneficial in the long run ;-)

jacoverster commented:
Ok, let's do that then, I also think it will be better.

rodja commented:
Strange. Now I got the build error:

```
14.09   building 'netifaces' extension
14.09   gcc -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -DNETIFACES_VERSION=0.11.0 -I/tmp/tmpb1la9hj8/.venv/include -I/usr/local/include/python3.8 -c netifaces.c -o build/temp.linux-aarch64-cpython-38/netifaces.o
14.09   In file included from /usr/lib/gcc/aarch64-linux-gnu/12/include/syslimits.h:7,
14.09                    from /usr/lib/gcc/aarch64-linux-gnu/12/include/limits.h:34,
14.09                    from /usr/local/include/python3.8/Python.h:11,
14.09                    from netifaces.c:1:
14.09   /usr/lib/gcc/aarch64-linux-gnu/12/include/limits.h:203:15: fatal error: limits.h: No such file or directory
14.09     203 | #include_next <limits.h>                /* recurse down to the real one */
14.09         |               ^~~~~~~~~~
14.09   compilation terminated.
14.09   error: command '/usr/bin/gcc' failed with exit code 1
14.09   
14.09 
14.09   at /usr/local/lib/python3.8/site-packages/poetry/installation/chef.py:147 in _prepare
14.10       143│ 
14.10       144│                 error = ChefBuildError("\n\n".join(message_parts))
14.10       145│ 
14.10       146│             if error is not None:
14.10     → 147│                 raise error from None
14.10       148│ 
14.10       149│             return path
14.10       150│ 
14.10       151│     def _prepare_sdist(self, archive: Path, destination: Path | None = None) -> Path:
14.10 
14.10 Note: This error originates from the build backend, and is likely not a problem with poetry but with netifaces (0.11.0) not supporting PEP 517 builds. You can verify this by running 'pip wheel --use-pep517 "netifaces (==0.11.0)"'.
14.10 
------
Dockerfile-with-features:32
--------------------
  30 |     RUN pip install -U pip && pip install poetry==$POETRY_VERSION
  31 |     COPY nicegui pyproject.toml poetry.lock README.md ./
  32 | >>> RUN poetry install --all-extras
  33 |     
  34 |     USER $USERNAME
--------------------
ERROR: failed to solve: process "/bin/sh -c poetry install --all-extras" did not complete successfully: exit code: 1
```

Seems `gcc` is not enough. But why has it worked with the multi-stage build? I can fix it by installing `build-essential` package instead.

rodja commented:
@natankeddem asked me if activating Rosetta emulation in Docker Desktop allows the pytests to run:

> @rodja does this help? https://github.com/browserless/chrome/issues/2004#issuecomment-1453092994

Unfortunately that does not change anything.

natankeddem commented:
https://askubuntu.com/a/1120102
That provides some ideas on the limit.h related dependencies.

rodja commented:
Ah, if I set `FROM --platform=linux/amd64 python:3.8-slim` in the Dockerfile the tests start and execute. But they often run into timeouts:

![Screenshot 2023-09-22 at 06 20 48](https://github.com/zauberzeug/nicegui/assets/131391/08e78887-1a80-429b-9858-f599d72b2940)

Looks like the emulation is not fast enough (even with Rosetta turned on).

jacoverster commented on .devcontainer/Dockerfile:
```diff
@@ -0,0 +1,35 @@
+ENV PATH="/home/${USERNAME}/.local/bin:${PATH}"
+
+# Install nicegui
+RUN pip install -U pip
```
Duplicate of next line, generally better to combine Run commands as far as possible

rodja replied:
Oh, sorry. I was not very attentive. I'll revert my commit!

jacoverster commented:
> Strange. Now I got the build error

This is strange. To further simplify things we can also consider moving away from the `slim` container. Maybe try that with the MacOS tests as well, it's possible that missing packages can cause the slow emulation time?

rodja commented:
> moving away from the slim container

If I remove the `-slim` suffix, it builds fine with `gcc`. But the tests still can't be executed when not using the amd64 image...

rodja commented:
I got the tests working on mac arm64 (M1). I just needed to configure the binary to chromium. See c8a6eee.

natankeddem commented:
So enabling rosetta and switching to amd64 did get things to work but too slowly it sounds like? That discussion on https://github.com/browserless/chrome/ also had some ideas on downloading and matching ARM driver versions. I haven't really dug in, but is that any help to you guys on getting M1 support working? It says to get drivers from here:
https://launchpad.net/ubuntu/bionic/arm64/chromium-chromedriver

rodja commented:
From my point of view we are almost ready (for a first "basic devcontainer"):
- documentation
- deciding if we want to use `build-essential` and `python.3.8-slim` (~100mb) or simply `python-3.8` which already comes with build-essential, wget curl, sed, awk and more useful development tools (~500mb); I'm leaning towards `python-3.8` without `-slim`.
- if I start the devcontainer and open a terminal it's still /bin/sh

jacoverster commented:
> * if I start the devcontainer and open a terminal it's still /bin/sh

I've compared built sizes here and `python-3.8` is at 2.19GB and `python.3.8-slim` is at 1.6GB for reference. No preference from my side, so happy with both.

> * if I start the devcontainer and open a terminal it's still /bin/sh

I've made another change to the settings in `devcontainer.json`, not sure what else to try here as mine defaults to `bash`. It is maybe worth trying to set this in your VSCode settings? Does that work?

> * documentation

I'm going on a few days of leave today. Will tackle this when I get back.

jacoverster commented:
One other images we can consider is the [official vscode one](https://github.com/devcontainers/images/tree/main/src/python). It has some nice benefits like built in remote user, security patches, multi-arch support, etc. Just putting it on the table as a potential option as well.

jacoverster commented:
> Interesting. I'll try it out.

Have a look at the linked [documentation ](https://containers.dev/guide/dockerfile) and the additional [features ](https://containers.dev/features )you can add to your container - might be something useful there for you.

natankeddem commented:
Should the extensions in `.devcontainer/devcontainer.json` match `nicegui.code-workspace`?

rodja commented:
> Should the extensions in .devcontainer/devcontainer.json match nicegui.code-workspace?

Yes that would be good!

rodja commented:
While the official python container from vscode looks interesting it may take some time and thinking to figure out how we want to do this. I would like to do this in a separate PR. This one is already quite complex and with 60+ comments a long time in the making.
Because bash now works for me also and documentation was quite easy to add, I consider the "basic version" to be ready for merge!

jacoverster commented:
Great, thanks for wrapping up the docs @rodja. Glad to get this merged.
